fetch/cache.go

```diff
@@ -15,20 +15,29 @@ import (
 // HashPeersCache holds lru cache of peers to pull hash from.
 type HashPeersCache struct {
 	*lru.Cache
-	mu    sync.Mutex
+	mu    sync.RWMutex
 	stats cacheStats
 }
 
 // HashPeers holds registered peers for a hash.
 type HashPeers map[p2p.Peer]struct{}
 
-// ToList converts hash peers map to a list.
-func (hp HashPeers) ToList() []p2p.Peer {
-	result := make([]p2p.Peer, 0, len(hp))
-	for k := range hp {
+// GetList returns hash peers as a list.
+func (hpc *HashPeersCache) GetList(hash types.Hash32) ([]p2p.Peer, bool) {
```

Comment:
is it possible to implement required api without copying? it is really bad pattern to copy in get, especially to copy a collection. better to use it only as a stop-gap, and definitely doesn't look good in newly written code

Reply:
another option is to store collection in the ordered map

Reply:
I think I’m missing something 🤔  `GetRandom` would need to convert peer map to a list too and then find a random peer there, right? What is a profit then? also I don’t quite understand how ordered map would help... how does it help to find a random element?